sql/core/src/main/scala/org/apache/spark/sql/expressions/ResolveVirtualColumn.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.expressions
+
+import org.apache.spark.sql.catalyst.analysis._
+import org.apache.spark.sql.catalyst.expressions.{Alias, VirtualColumn}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.expressions.SparkPartitionID
+
+/**
+ * Used to resolve virtual column attributes and insert the expression for them. Add case
+ * statements to this for each additional virtual column
+ */
+private[sql] class ResolveVirtualColumn(resolver: Resolver) extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case lPlan: LogicalPlan => lPlan transformExpressionsUp {
+      case u @ UnresolvedAttribute(nameParts) if nameParts.length == 1 &&
+        resolver(nameParts.head, VirtualColumn.sparkPartitionId) => SparkPartitionID
```

Comment:
So the query you gave will resolve the outer `spark__partition__id` as a virtual column, ignoring the alias. 

Some other data points:

`select spark__partition__id from table` (where `spark__partition__id` is a column) will resolve to the column and not the virtual column.

`SELECT a as spark__partition__id from test_table` will resolve the column `a`

The order of precedence goes, column, virtual_column, alias.